oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentials.java

```diff
@@ -0,0 +1,244 @@
+        credentialSourceType = IdentityPoolCredentialSourceType.FILE;
+      } else {
+        credentialLocation = (String) credentialSourceMap.get("url");
+        credentialSourceType = IdentityPoolCredentialSourceType.URL;
+      }
+
+      Map<String, String> headersMap = (Map<String, String>) credentialSourceMap.get("headers");
+      if (headersMap != null && !headersMap.isEmpty()) {
+        headers = new HashMap<>();
+        headers.putAll(headersMap);
+      }
+    }
+
+    private boolean hasHeaders() {
+      return headers != null && !headers.isEmpty();
+    }
+  }
+
+  /**
+   * Internal constructor. See {@link
+   * ExternalAccountCredentials#ExternalAccountCredentials(HttpTransportFactory, String, String,
+   * String, String, CredentialSource, String, String, String, String, Collection)}
+   */
+  IdentityPoolCredentials(
+      HttpTransportFactory transportFactory,
+      String audience,
+      String subjectTokenType,
+      String tokenUrl,
+      String tokenInfoUrl,
+      IdentityPoolCredentialSource credentialSource,
+      @Nullable String serviceAccountImpersonationUrl,
+      @Nullable String quotaProjectId,
+      @Nullable String clientId,
+      @Nullable String clientSecret,
+      @Nullable Collection<String> scopes) {
+    super(
+        transportFactory,
+        audience,
+        subjectTokenType,
+        tokenUrl,
+        tokenInfoUrl,
+        credentialSource,
+        serviceAccountImpersonationUrl,
+        quotaProjectId,
+        clientId,
+        clientSecret,
+        scopes);
+  }
+
+  @Override
+  public AccessToken refreshAccessToken() throws IOException {
+    String credential = retrieveSubjectToken();
+    StsTokenExchangeRequest.Builder stsTokenExchangeRequest =
+        StsTokenExchangeRequest.newBuilder(credential, subjectTokenType).setAudience(audience);
+
+    if (scopes != null && !scopes.isEmpty()) {
+      stsTokenExchangeRequest.setScopes(new ArrayList<>(scopes));
+    }
+
+    AccessToken accessToken = exchange3PICredentialForAccessToken(stsTokenExchangeRequest.build());
+    return attemptServiceAccountImpersonation(accessToken);
+  }
+
+  @Override
+  public String retrieveSubjectToken() throws IOException {
+    IdentityPoolCredentialSource identityPoolCredentialSource =
+        (IdentityPoolCredentialSource) credentialSource;
+    if (identityPoolCredentialSource.credentialSourceType
+        == IdentityPoolCredentialSource.IdentityPoolCredentialSourceType.FILE) {
+      return retrieveSubjectTokenFromCredentialFile();
+    }
+    return getSubjectTokenFromMetadataServer();
+  }
+
+  private String retrieveSubjectTokenFromCredentialFile() throws IOException {
+    IdentityPoolCredentialSource identityPoolCredentialSource =
+        (IdentityPoolCredentialSource) credentialSource;
+    String credentialFilePath = identityPoolCredentialSource.credentialLocation;
+    if (!new File(credentialFilePath).isFile()) {
```

Comment:
Files.exists instead of a new file object

Reply:
Done.